### PR TITLE
feat(std/node/fs): adding some functions

### PR DESCRIPTION
**Pull request by AliBasicCoder**
this pull request add all messing functions in `std/node/fs` that could be added
but it not complete yet and i need some help

list of already added functions:
1. readdir
1. readdirSync
1. rename
1. renameSync
1. rmdir
1. rmdirSync
1. unlink
1. unlinkSync
1. lstat
1. lstatSync
1. stat
1. statSync

### CODE REVIEW

**Comment by bartlomieju** [comment]
Thanks for the PR @AliBasicCoder. Make sure to properly rebase the branch as your PR currently contains a lot of unrelated code to this PR.

**Review by crypt096 — APPROVED** [review]

**Comment by AliBasicCoder** [comment]
pull request have been rebased

**Comment by AliBasicCoder** [comment]
how to add `fs.symlink` and `fs.symlinkSync` while `Deno.symlink` is not stable?

**Comment by AliBasicCoder** [comment]
what is the goal of std/node
is it having 100% of node.js api
or just help some one who came from node to deno get started?

**Comment by AliBasicCoder** [comment]
the PR is properly rebased now

**Comment by AliBasicCoder** [comment]
this PR is ready for merging

**Comment by AliBasicCoder** [comment]
> Thanks for the PR @AliBasicCoder. Make sure to properly rebase the branch as your PR currently contains a lot of unrelated code to this PR.

the PR is rebased and ready for merging

**Comment by AliBasicCoder** [comment]
@bartlomieju please review this PR

File: std/node/_fs/_fs_open.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
+import { existsSync } from "../../fs/mod.ts";
+import { fromFileUrl } from "../path.ts";
+
+type openFlags =
+  | "a"
+  | "ax"
+  | "a+"
+  | "ax+"
+  // | "as"
+  // | "as+"
```
Comment (bartlomieju) [inline std/node/_fs/_fs_open.ts:10]: Why are those commented out?

File: std/node/_fs/_fs_readdir.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+export function readdir(path: string | URL, callback: readDirCallback): void;
+export function readdir(
+  path: string | URL,
+  optionsOrCallback: readDirOptions | readDirCallback,
+  maybeCallback?: readDirCallback,
+) {
+  const callback = typeof optionsOrCallback === "function"
+    ? optionsOrCallback
+    : maybeCallback;
+  // const options = typeof optionsOrCallback === "object" ? optionsOrCallback : null;
```
Comment (bartlomieju) [inline std/node/_fs/_fs_readdir.ts:24]: remove?

File: std/node/_fs/_fs_readdir.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      }
+      result.push(val.name);
+    });
+  } catch (error) {
+    callback(error, result);
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export function readdirSync(path: string | URL, options?: readDirOptions) {
```
Comment (bartlomieju) [inline std/node/_fs/_fs_readdir.ts:43]: why are options unused?

File: std/node/_fs/_fs_stat_test.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  assertEquals(actual.mtime?.getTime(), expected.mtime?.getTime());
+  assertEquals(actual.birthtime?.getTime(), expected.birthtime?.getTime());
+  assertEquals(actual.atimeMs, expected.atime?.getTime());
+  assertEquals(actual.mtimeMs, expected.mtime?.getTime());
+  assertEquals(actual.birthtimeMs, expected.birthtime?.getTime());
+  assertEquals(actual.isFile(), expected.isFile);
+  assertEquals(actual.isDirectory(), expected.isDirectory);
+  assertEquals(actual.isSymbolicLink(), expected.isSymlink);
+  // assertEquals(actual.ctimeMs === expected.ctime?.getTime());
+  // assertEquals(actual.ctime?.getTime() === expected.ctime?.getTime());
```
Comment (bartlomieju) [inline std/node/_fs/_fs_stat_test.ts:26]: Remove?

File: std/node/_fs/_fs_stat_test.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  assertEquals(Number(actual.birthtimeMs), expected.birthtime?.getTime());
+  assertEquals(Number(actual.atimeNs) / 1e6, expected.atime?.getTime());
+  assertEquals(Number(actual.mtimeNs) / 1e6, expected.atime?.getTime());
+  assertEquals(Number(actual.birthtimeNs) / 1e6, expected.atime?.getTime());
+  assertEquals(actual.isFile(), expected.isFile);
+  assertEquals(actual.isDirectory(), expected.isDirectory);
+  assertEquals(actual.isSymbolicLink(), expected.isSymlink);
+  // assertEquals(actual.ctime?.getTime() === expected.ctime?.getTime());
+  // assertEquals(Number(actual.ctimeMs) === expected.ctime?.getTime());
+  // assertEquals(Number(actual.ctimeNs) / 1e+6 === expected.ctime?.getTime());
```
Comment (bartlomieju) [inline std/node/_fs/_fs_stat_test.ts:58]: Ditto

File: std/node/fs.ts [hunk 6]
```diff
@@ -11,6 +11,16 @@ import { exists, existsSync } from "./_fs/_fs_exists.ts";
 import { mkdir, mkdirSync } from "./_fs/_fs_mkdir.ts";
 import { copyFile, copyFileSync } from "./_fs/_fs_copy.ts";
 import { writeFile, writeFileSync } from "./_fs/_fs_writeFile.ts";
+import { readdir, readdirSync } from "./_fs/_fs_readdir.ts";
+import { rename, renameSync } from "./_fs/_fs_rename.ts";
+import { rmdir, rmdirSync } from "./_fs/_fs_rmdir.ts";
+import { unlink, unlinkSync } from "./_fs/_fs_unlink.ts";
+import { watch } from "./_fs/_fs_watch.ts";
+import { open, openSync } from "./_fs/_fs_open.ts";
+// not sure about those
```
Comment (bartlomieju) [inline std/node/fs.ts:20]: Why?

**Review by bartlomieju — COMMENTED** [review]
@AliBasicCoder thanks! Great PR, some questions before it's good to land

File: std/node/_fs/_fs_open.ts (hunk 1 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_open.ts:10]: well, in the [node docs](https://nodejs.org/api/fs.html#fs_file_system_flags)
it says ``'as': Open file for appending in synchronous mode. The file is created if it does not exist.``
i don't know if using "as" and "as+" flags should make fs.open synchronous or not?
but i tested in node and it's not sync

File: std/node/_fs/_fs_readdir.ts (hunk 2 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_readdir.ts:24]: options could be used should I?

File: std/node/_fs/_fs_stat_test.ts (hunk 4 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_stat_test.ts:26]: Deno doesn't provide a "ctime" in stat and lstat so, i used "mtime" for "ctime"

File: std/node/fs.ts (hunk 6 above)
Comment (AliBasicCoder) [reply std/node/fs.ts:20]: it's an old comment that i forgot to delete sorry

File: std/node/_fs/_fs_open.ts (hunk 1 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_open.ts:10]: i can add them if you want

File: std/node/_fs/_fs_readdir.ts (hunk 3 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_readdir.ts:43]: should i add them??

**Review by james-and — COMMENTED** [review]
ok im getting there

File: std/node/_fs/_fs_open.ts (hunk 1 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_open.ts:10]: > Why are those commented out? 

well, in the [node docs](https://nodejs.org/api/fs.html#fs_file_system_flags)
it says ``'as': Open file for appending in synchronous mode. The file is created if it does not exist.``
i don't know if using "as" and "as+" flags should make fs.open synchronous or not?

File: std/node/_fs/_fs_stat_test.ts (hunk 4 above)
Comment (AliBasicCoder) [reply std/node/_fs/_fs_stat_test.ts:26]: > Remove?

yes

File: std/node/fs.ts (hunk 6 above)
Comment (AliBasicCoder) [reply std/node/fs.ts:20]: > Why?

it's an old comment that i forgot to delete sorry

**Comment by AliBasicCoder** [comment]
@bartlomieju l have a commit for using `open` and `openSync` options and another one for using `as` and `as+` flag in sync mode on my local machine what should i push it??

**Comment by AliBasicCoder** [comment]
so i did the following:
- handled options in readdir and readdirSync
- handled flags "as" and "as+" in open and openSync
- fix issue with watch

**Comment by AliBasicCoder** [comment]
@ry

**Review by ry — APPROVED** [review]
LGTM - nice tests

**Pull request merged**